### PR TITLE
[Telemetry] Use CounterMetadata subclasses

### PR DESCRIPTION
**Pull request by iainx**
Use CounterMetadata subclasses so that metadata properties aren't reduced to
strings and will be placed into the correct property or measures bag

Fixes VSTS #633544

### CODE REVIEW

**Comment by iainx** [comment]
Related to https://github.com/xamarin/md-addins/pull/3261

File: main/src/core/MonoDevelop.Core/MonoDevelop.Core.Instrumentation/TimerCounter.cs [hunk 1]
```diff
@@ -213,55 +213,42 @@ public ITimeTracker<T> BeginTiming (string message, T metadata, CancellationToke
...
 
-		protected void SetProperty (object value, [CallerMemberName]string name = null)
-			=> properties[name] = Convert.ToString (value, CultureInfo.InvariantCulture);
-
-		protected string GetProperty ([CallerMemberName]string name = null)
+		protected T GetProperty<T> ([CallerMemberName]string name = null)
 		{
 			properties.TryGetValue (name, out var result);
-			return result;
+			return (result is T) ? (T)result : default(T);
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Core.Instrumentation/TimerCounter.cs]: This is going to hide bugs. I'd rather have exceptions here.

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/ProjectService.cs [hunk 2]
```diff
@@ -477,23 +464,77 @@ internal static class Counters
...
+		public ReadSolutionItemMetadata ()
+		{
+		}
+
+		public ReadSolutionItemMetadata (string fileNameExtension, string typeGuid, string itemGuid)
+		{
+			SetProperty (fileNameExtension, "FileNameExtension");
+			SetProperty (false, "LoadSucceed");
+			if (typeGuid != null) {
+				SetProperty (typeGuid, "ProjectType");
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects/ProjectService.cs]: `ProjectType = typeGuid`? same with above

File: main/src/core/MonoDevelop.Core/MonoDevelop.Core.Instrumentation/Counter.cs [hunk 3]
```diff
@@ -271,6 +278,12 @@ public void Inc (IDictionary<string, string> metadata)
 		}
 
 		public void Inc (int n, string message, IDictionary<string, string> metadata)
```
Comment (slluis) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Core.Instrumentation/Counter.cs]: Please mark this method as obsolete, and instruct to use the overload that takes `IDictionary<string, object>` instead. Same for all other methods in this class that take `IDictionary<string, string>`

File: main/src/core/MonoDevelop.Core/MonoDevelop.Core.Instrumentation/TimerCounter.cs [hunk 4]
```diff
@@ -117,12 +117,12 @@ public ITimeTracker BeginTiming (string message)
 			return BeginTiming (message, null);
 		}
 
-		public ITimeTracker BeginTiming (IDictionary<string, string> metadata)
+		public ITimeTracker BeginTiming (IDictionary<string, object> metadata)
```
Comment (slluis) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Core.Instrumentation/TimerCounter.cs]: This is an API break. Add a new overload and mark the old one as Obsolete.

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Navigation/Counters.cs [hunk 5]
```diff
@@ -27,31 +27,29 @@
...
-			var metadata = new Dictionary<string, string> ();
-			metadata ["Type"] = navigationType;
-			metadata ["Result"] = "Failure"; // Will be updated when navigation finishes.
+			var metadata = new NavigationMetadata (navigationType);
+			metadata.SetResult (false);
 			return metadata;
 		}
 
-		public static void UpdateUserCancellation (IDictionary<string, string> metadata, CancellationToken cancellationToken)
+		public static void UpdateUserCancellation (NavigationMetadata metadata, CancellationToken cancellationToken)
```
Comment (slluis) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.Navigation/Counters.cs]: Please remove this method and replace with calls to `NavigationMetadata.UpdateUserCancellation`.

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Navigation/Counters.cs [hunk 6]
```diff
@@ -27,31 +27,29 @@
...
 		{
-			if (cancellationToken.IsCancellationRequested) {
-				metadata ["Result"] = "UserCancel";
-			}
+			metadata.UpdateUserCancellation (cancellationToken);
 		}
 
-		public static void UpdateNavigateResult (IDictionary<string, string> metadata, bool result)
+		public static void UpdateNavigateResult (NavigationMetadata metadata, bool result)
```
Comment (slluis) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.Navigation/Counters.cs]: Please remove this method and replace with calls to `NavigationMetadata.SetResult`.

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Navigation/Counters.cs [hunk 7]
```diff
@@ -61,9 +59,9 @@ public static void UpdateNavigateResult (IDictionary<string, string> metadata, b
 		/// will not attempt to find any extension methods. Note that the CommandInfo is disabled for
 		/// the menu but this does not seem to have any affect presumably because the method is async.
 		/// </summary>
-		public static void UpdateUserFault (IDictionary<string, string> metadata)
+		public static void UpdateUserFault (CounterMetadata metadata)
```
Comment (slluis) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.Navigation/Counters.cs]: Please remove this method and replace with calls to `NavigationMetadata.SetUserFault`.

File: main/src/addins/MonoDevelop.Refactoring/MonoDevelop.Refactoring/RefactoringService.cs [hunk 8]
```diff
@@ -354,23 +354,32 @@ public static MonoDevelop.Ide.Editor.DocumentLocation GetCorrectResolveLocation
...
-			var metadata = new Dictionary<string, string> ();
-			metadata ["Result"] = "Success";
+			var metadata = new FindReferencesMetadata {
+				ResultString = "Success"
+			};
 			return metadata;
 		}
 
-		public static void SetFailure (IDictionary<string, string> metadata)
+		public static void SetFailure (FindReferencesMetadata metadata)
```
Comment (slluis) [inline main/src/addins/MonoDevelop.Refactoring/MonoDevelop.Refactoring/RefactoringService.cs]: Handle status like in `NavigationMetadata`. Remove all static methods. All logic must be inside the CounterMetadata subclass.

File: main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/Repository.cs [hunk 9]
```diff
@@ -42,10 +42,11 @@ protected Repository ()
 		protected Repository (VersionControlSystem vcs): this ()
 		{
 			VersionControlSystem = vcs;
-			Repositories.SetValue (Repositories.Count + 1, string.Format ("Repository #{0}", Repositories.Count + 1), new Dictionary<string, string> {
-				{ "Type", vcs.Name },
-				{ "Type+Version", string.Format ("{0} {1}", vcs.Name, vcs.Version) },
-			});
+			var metadata = new RepositoryMetadata {
+				Type = vcs.Name,
+				Version = vcs.Version
```
Comment (slluis) [inline main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/Repository.cs]: Please don't change how telemetry is reported. Bring back the "Type+Version" property.

File: main/src/core/MonoDevelop.Core/MonoDevelop.Core.Instrumentation/TimerCounter.cs [hunk 10]
```diff
@@ -213,55 +213,42 @@ public ITimeTracker<T> BeginTiming (string message, T metadata, CancellationToke
...
-					return CounterResult.Unspecified;
-				return res;
-			}
-			set {
-				if (value == CounterResult.Unspecified)
-					properties.Remove (nameof (Result));
-				else
-					SetProperty (value.ToString ());
-			}
+		public string Result {
```
Comment (slluis) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Core.Instrumentation/TimerCounter.cs]: Why change this to a string?

File: main/src/core/MonoDevelop.Core/MonoDevelop.Core.Instrumentation/TimerCounter.cs [hunk 11]
```diff
@@ -213,55 +213,42 @@ public ITimeTracker<T> BeginTiming (string message, T metadata, CancellationToke
...
-		public void SetFailure () => Result = CounterResult.Failure;
-		public void SetSuccess () => Result = CounterResult.Success;
-		public void SetUserCancel () => Result = CounterResult.UserCancel;
+		public void SetUserFault () => Result = CounterResult.UserFault.ToString ();
+		public void SetFailure () => Result = CounterResult.Failure.ToString ();
+		public void SetSuccess () => Result = CounterResult.Success.ToString ();
+		public void SetUserCancel () => Result = CounterResult.UserCancel.ToString ();
 
-		protected void SetProperty (string value, [CallerMemberName]string name = null)
+		public void SetProperty (object value, [CallerMemberName]string name = null)
```
Comment (slluis) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Core.Instrumentation/TimerCounter.cs]: Don't change to public.

File: main/src/core/MonoDevelop.Core/MonoDevelop.Core.Instrumentation/TimerCounter.cs [hunk 12]
```diff
@@ -213,55 +213,42 @@ public ITimeTracker<T> BeginTiming (string message, T metadata, CancellationToke
...
-		protected string GetProperty ([CallerMemberName]string name = null)
+		protected T GetProperty<T> ([CallerMemberName]string name = null)
 		{
 			properties.TryGetValue (name, out var result);
-			return result;
+			return (result is T) ? (T)result : default(T);
 		}
 
-		protected T GetProperty<T> ([CallerMemberName]string name = null)
+		public bool ContainsProperty (string propName)
```
Comment (slluis) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Core.Instrumentation/TimerCounter.cs]: Make this protected.

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/Project.cs [hunk 13]
```diff
@@ -746,7 +746,7 @@ internal protected override IEnumerable<string> GetItemTypeGuids ()
 			return base.GetItemTypeGuids ().Concat (flavorGuids);
 		}
 
-		protected override void OnGetProjectEventMetadata (IDictionary<string, string> metadata)
+		protected override void OnGetProjectEventMetadata (ProjectEventMetadata metadata)
```
Comment (slluis) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects/Project.cs]: You can't change this, it is a public API break.

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/Project.cs [hunk 14]
```diff
@@ -1407,27 +1407,27 @@ internal protected virtual Task<TargetEvaluationResult> OnRunTarget (ProgressMon
...
+				metadata.BuildType = 1;
 			}
-			metadata ["BuildTypeString"] = target;
+			metadata.BuildTypeString = target;
 
-			metadata ["FirstBuild"] = IsFirstBuild.ToString ();
-			metadata ["ProjectID"] = ItemId;
-			metadata ["ProjectType"] = TypeGuid;
-			metadata ["ProjectFlavor"] = FlavorGuids.FirstOrDefault () ?? TypeGuid;
+			metadata.FirstBuild = IsFirstBuild.ToString ();
```
Comment (slluis) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects/Project.cs]: No need to convert to string anymore.

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/SolutionItem.cs [hunk 15]
```diff
@@ -866,21 +866,22 @@ protected virtual Task<BuildResult> OnClean (ProgressMonitor monitor, Configurat
 			inserted[index] = true;
 		}
 		
-		public IDictionary<string, string> GetProjectEventMetadata (ConfigurationSelector configurationSelector)
+		public ProjectEventMetadata GetProjectEventMetadata (ConfigurationSelector configurationSelector)
```
Comment (slluis) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects/SolutionItem.cs]: API break.

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/SolutionItem.cs [hunk 16]
```diff
@@ -866,21 +866,22 @@ protected virtual Task<BuildResult> OnClean (ProgressMonitor monitor, Configurat
...
 
-			OnGetProjectEventMetadata (data);
-			return data;
+			var metadata = new ProjectEventMetadata (id);
+			OnGetProjectEventMetadata (metadata);
+			return metadata;
 		}
 
-		protected virtual void OnGetProjectEventMetadata (IDictionary<string, string> metadata)
+		protected virtual void OnGetProjectEventMetadata (ProjectEventMetadata metadata)
```
Comment (slluis) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects/SolutionItem.cs]: API break.

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/SolutionItem.cs [hunk 17]
```diff
@@ -1783,4 +1784,73 @@ public override object GetTagValue (SolutionItem item, string tag)
...
 		}
 	}
+
+	public class ProjectEventMetadata : CounterMetadata
+	{
+		public ProjectEventMetadata ()
+		{
+		}
+
+		public ProjectEventMetadata (string id)
```
Comment (slluis) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects/SolutionItem.cs]: Rename argument to configurationId

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Templates/MicrosoftTemplateEngineProjectTemplatingProvider.cs [hunk 18]
```diff
@@ -108,11 +109,12 @@ static internal SolutionTemplate CreateTemplate (string templateId, string scanP
...
-			var metadata = new Dictionary<string, string> ();
-			metadata ["Id"] = templateInfo.Identity;
-			metadata ["Name"] = templateInfo.Name;
-			metadata ["Language"] = template.Language;
-			metadata ["Platform"] = string.Join(";", templateInfo.Classifications);
+			var metadata = new TemplateMetadata {
+				Id = templateInfo.Identity,
+				Name = templateInfo.Name,
+				Language = template.Language,
+				Platform = string.Join("l", templateInfo.Classifications)
```
Comment (slluis) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Templates/MicrosoftTemplateEngineProjectTemplatingProvider.cs]: Join with "l"???

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/RootWorkspace.cs [hunk 19]
```diff
@@ -1541,6 +1541,39 @@ public void Dispose ()
...
+			get => GetProperty<bool> ();
+			set => SetProperty (value);
+		}
+
+		public bool LoadSucceed {
+			get => GetProperty<bool> ();
+			set => SetProperty (value);
+		}
+
+		public string Reason {
```
Comment (slluis) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/RootWorkspace.cs]: Use an enum for the reason.

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/SolutionItem.cs [hunk 20]
```diff
@@ -1783,4 +1784,73 @@ public override object GetTagValue (SolutionItem item, string tag)
...
+			set => SetProperty (value);
+		}
+
+		public string ProjectFlavor {
+			get => GetProperty<string> ();
+			set => SetProperty (value);
+		}
+
+		public string Configuration {
+			get => GetProperty<string> ();
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects/SolutionItem.cs]: Should this map to `Config.Id`? If so, explicitly set it.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Templates/MicrosoftTemplateEngineProjectTemplatingProvider.cs (hunk 18 above)
Comment (Therzok) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Templates/MicrosoftTemplateEngineProjectTemplatingProvider.cs]: Should be `;`

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Navigation/Counters.cs [hunk 21]
```diff
@@ -75,23 +73,27 @@ public NavigationMetadata ()
...
 			{
 				Properties["Type"] = type;
-				Result = CounterResult.Failure; // Will be updated when navigation finishes.
+				SetFailure ();// Will be updated when navigation finishes.
 			}
 
 			public string Type {
-				get { return Properties.TryGetValue ("Type", out var type) ? type : null; }
-				set { Properties ["Type"] = value; }
+				get => GetProperty<string> ();
```
Comment (Therzok) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.Navigation/Counters.cs]: `Properties.ContainsKey ("Type") ? GetProperty<string>() : null`

File: main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/Repository.cs (hunk 9 above)
Comment (iainx) [reply main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/Repository.cs]: I didn't see that the `Type+Version` property was actually being reported anywhere so I thought we could split them up for working with the data easier. I'll rejoin them, but `+` isn't allowed in a telemetry name so I'll just call it TypeVersion

File: main/src/core/MonoDevelop.Core/MonoDevelop.Core.Instrumentation/Counter.cs (hunk 3 above)
Comment (iainx) [reply main/src/core/MonoDevelop.Core/MonoDevelop.Core.Instrumentation/Counter.cs]: Would it be better to obsolete the whole class and use the generic version instead? Offering a public API with `<string, object>` doesn't seem very nice to me

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/SolutionItem.cs (hunk 20 above)
Comment (iainx) [reply main/src/core/MonoDevelop.Core/MonoDevelop.Projects/SolutionItem.cs]: No, they were two separate properties in the old code:

https://github.com/mono/monodevelop/pull/5093/files/c3e75f3ef6487c47ac46bec85171c47e4e7df042#diff-a11e7d82eb58a33105c896fbbe686b3bL875 vs

https://github.com/mono/monodevelop/pull/5093/files/c3e75f3ef6487c47ac46bec85171c47e4e7df042#diff-1728662d3af95d7e7373087b558c05feL1429

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Editor.Extension/CompletionStatistics.cs [hunk 22]
```diff
@@ -83,18 +84,61 @@ public void Report ()
...
+				FailureCount = failureCount,
+				SuccessCount = successCount,
+				CancelCount = cancelCount
+			};
 
 			Counters.CodeCompletionStats.Inc (metadata);
 		}
 	}
+
+	public class CompletionStatisticsMetadata : CounterMetadata
```
Comment (slluis) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Editor.Extension/CompletionStatistics.cs]: Please make private.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui/Workbench.cs [hunk 23]
```diff
@@ -1651,4 +1653,33 @@ public enum OpenDocumentOptions
 		Debugger = BringToFront | CenterCaretLine | TryToReuseViewer,
 		DefaultInternal = Default | OnlyInternalViewer,
 	}
+
+	public class OpenDocumentMetadata : CounterMetadata
```
Comment (slluis) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui/Workbench.cs]: Should not be public.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Templates/MicrosoftTemplateEngineProjectTemplatingProvider.cs [hunk 24]
```diff
@@ -187,4 +189,31 @@ static IEnumerable<TemplateParameter> GetValidParameters (string parameters)
 				.Where (parameter => parameter.IsValid);
 		}
 	}
+
+	public class TemplateMetadata : CounterMetadata
```
Comment (slluis) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Templates/MicrosoftTemplateEngineProjectTemplatingProvider.cs]: Should not be public.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/RootWorkspace.cs [hunk 25]
```diff
@@ -1545,6 +1545,55 @@ public void Dispose ()
 			NotifyChanges ();
 		}
 	}
+
+
+	public class OpenWorkspaceItemMetadata : CounterMetadata
```
Comment (slluis) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/RootWorkspace.cs]: Should not be public.

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/Project.cs [hunk 26]
```diff
@@ -715,7 +717,23 @@ protected override void OnGetProjectEventMetadata (IDictionary<string, string> m
 				sb.Append (p);
 				first = false;
 			}
-			metadata ["ProjectTypes"] = sb.ToString ();
+			metadata["ProjectTypes"] = sb.ToString ();
+		}
+
+		protected override void OnGetProjectEventMetadata (ProjectEventMetadata metadata)
```
[the remark this answers is not in the document]
Comment (slluis) [reply main/src/core/MonoDevelop.Core/MonoDevelop.Projects/Project.cs]: I see, let's keep it for now.

**Comment by slluis** [comment]
@iainx a few tests are failing, they need to be adapted to the api changes.

**Comment by slluis** [comment]
@Therzok can you update your review?

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/SolutionItem.cs [hunk 27]
```diff
@@ -865,24 +865,54 @@ protected virtual Task<BuildResult> OnClean (ProgressMonitor monitor, Configurat
...
+					id = slnConfig.Id;
+				}
+			}
+
+			var metadata = new ProjectEventMetadata (id);
+			OnGetProjectEventMetadata (metadata);
+			return metadata;
+		}
+
+		public void UpdateProjectEventMetadata (ProjectEventMetadata metadata)
```
Comment (slluis) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects/SolutionItem.cs]: This method is not necessary anymore.

**Comment by Therzok** [comment]
A RE bump was accidentally added.

**Comment by iainx** [comment]
*sigh*

**Pull request merged**